### ISSUE TITLE
Different access modifier for getter and setter

### ISSUE THREAD

**Issue by Andree643** [microsoft/TypeScript#2845]
Could it be possible to implement different access modifier for getter and setter? This way the setter could be for example private/protected and the getter public. In some cases this in really useful when the value shall be read-only.

Example:

```
class MyClass {
    private _myProp: any;
    private set myProp(value: any) {
        this._myProp = value;
    }
    public get myProp(): any {
        return this._myProp;
    }
}
```

**Comment by RyanCavanaugh** [comment]
#12

**Comment by raveclassic** [comment]
As long as #12 is closed with #6532 now it seems that this issue is out of scope now.
Do you now have any plans on implementing different access modifiers? 
Because `readonly` is not enough to solve what's described here as a property might be actually writable inside a class.

**Comment by raveclassic** [comment]
I believe I should also move my example here from related issue:
```typescript
declare abstract class Emitter {
	new (): Emitter;
	on: (name:string, handler: (arg:any) => void) => void;
	off: (name:string, handler: (arg:any) => void) => void;
	protected emit: (name:string, arg:any) => void;
}

class Person extends Emitter {
	constructor(name:string) {
		super();
		this.name = name;
	}
	
	private _name:string;
	get name() {
		return this._name;
	}
	set name(value) {
		if (this._name !== value) {
			this._name = value;
			this.emit('change:name', value);
			//this way is better
			this.updatedAt = new Date();
			//than this way
			this.setUpdatedAt(new Date());
		}
	}
	
	////
	private _updatedAt:Date;
	get updatedAt() {
		return this._updatedAt;
	}
	private set updatedAt(value) { //Getter and setter do not agree in visibility
                //some logic and a simple readonly (our absence of a setter) is not enough
		if (this._updatedAt !== value) {
			this._updatedAt = value;
			this.emit('change:updatedAt', value);
		}
	}
	
	//// method implementation - but what's the point in it?
	private setUpdatedAt(value) {
		if (this._updatedAt !== value) {
			this._updatedAt = value;
			this.emit('change:updatedAt', value);
		}
	}
}

const entity = new Person('Mike');
entity.on('change:updatedAt', console.log.bind(console));
entity.name = 'Thomas';
//but manually setting updatedAt should be forbidden
entity.updatedAt = new Date(); //restricted
```

Here, property `updatedAt` actually **can** have a setter but it **should not be accessable** outside of  `Person`. Moreover this setter contains complex logic and a simple readonly or an absence of a setter is not enough. 

I agree that private setter is just a syntax sugar for private methods with additional logic (like emitting) but I think it is inconsistent when part of logic related to a field is in a property (getter) and another in a method (actual private setter).

This way getters/setter are implemented in `C#`, I think it would be usefull to allow different visibility at a compile time, though this contracts will be lost in runtime.

**Comment by mhegazy** [comment]
The recommendation here is to use a public readonly getter and a private/protected backing field. 

Accessors are symetrical with properties in the type system. anything we do will need to be manifested in the type and expressible on properties. Adding new access modifiers to enable private_set/public_get would increase the complexity of the language and the learning curve, and the value gained from this would not match the added complexity.

**Comment by 0815fox** [comment]
Anyway adding my +1 for this, in case it is considered again in future...

**Comment by Roam-Cooper** [comment]
I'd still like to see this. C# has it, and it's incredibly useful in many circumstances, especially when you have flags that want to be read externally to the class, but only set internally with private/protected.

I think the suggestion that it isn't used all that often is a farce in that the pattern isn't used because it's not available to use.

**Comment by zommarin** [comment]
This is a good addition to the language: The amount of complexity that this would add (from the perspective of a typescript programmer) is small. The concept is easy to understand and the compiler should be able to provide good error messages hinting of why you cannot access getters or setters due to scoping.

**Comment by mjohnk11** [comment]
I agree with the above users, this is standard for most other languages. The argument against it does not hold weight and the implementation should be reconsidered.

**Comment by blteblte** [comment]
I actually was surprised you can not do this... +1 to the issue.
There shouldn't bee any increased learning curve with this
```
private get x() { ... }
public set x(value) { ... }
```
Imo if you can read english it is self-explanatory what private(protected)/public mean here. Also if you are defining the accessors in a first place - you probably already know what they are for.

**Comment by ianfp** [comment]
Here are two use cases where this would be handy:

Backbone.js, to avoid ugly `.get()` and `.set()` calls:
```typescript
class Whatever {
    public get rotation(): number {
        return this.get('rotation');
    }
    private set rotation(rotation: number) {
        this.set('rotation', rotation);
    }
}
```

Properties that subclasses can modify:
```typescript
class ExtendMe {
    public get someProperty(): string {
        // some stuff
    }
    protected set someProperty(prop: string) {
        // some stuff
    }
}
```

**Comment by ZLima12** [comment]
I definitely would like to see this, as it has been bugging me since I started using TypeScript.

**Comment by marc-gj** [comment]
I agree with the people requesting this feature. I just tried having a public get() method and was surprised to see that my set and get must agree on visibility.

**Comment by mdarefull** [comment]
I've been reading you guys claiming it is too complex. Remembering the "C++ way" why it is different from:
```
private _myAttribute: string;
get myAttribute(): string {...}
setMyAttribute(value: string) {...}
```

I can see a lot of use cases for this, that's why I'm here right now. 
What if I want myAttribute to be publicly accessible but allow only to be modified inside its class? 
What if I want to add some custom logic each time the attribute is modified? 
It can be a business rule, it can be just a logging to understand why it has been assigned a specific value, together with a breakpoint condition for debugging purposes, etc...

Basically, we want a method to be used each time we modify the attribute together with the syntactic sugar that makes it seems we're just assigning a value, instead than calling a method.
C# has the concept of properties for this and I thought TS inherited that concept, but not allowing different accessibilities is a big limitations for people that think like me and make us fall back to the "C++ style".

**Comment by Giwayume** [comment]
"It's too hard" should never be a reason to not implement an incredibly useful feature.

**Comment by tempestwf** [comment]
Adding my +1. It seems very odd to me that in a language that is so completely in so many regards you can't do this.

**Comment by skynode** [comment]
Surprised to find that this issue is marked as closed. This feature is DESIRED by the community. Please reopen. 
+1

**Comment by TomasHubelbauer** [comment]
#12 doesn't address this issue, so this issue should not be closed with a comment that refers to that issue.

**Comment by iamjoyce** [comment]
Not saying this feature isn't necessary but here's a way to bypass for `set`. IMO `get` should always return the same variable type, so this was good enough for me.

```
class Widget {
    get price(): number { return this._price; }
    set price(val){ return this.setPrice(val); } // call another function

    // do processing here
    private setPrice(price: number | string): number {
        let num = Number(price);
        return isNaN(num) ? 0 : num;
    }
}

```

**Comment by idchlife** [comment]
IMHO this is a nice feature but as fair as I understand it's just the situation "well I do NOT wan't to write everytime _property instead I want to use private set property"

If there is choice either adding huge complexity to the TypeScript core code itself (remember, more unnecessary complexity - less speed of developing things, more time to debug, write tests, in the end - less frequent releases) but reducing writing 1 symbol OR writing that 1 symbol and deal with it well... I would rather write 1 symbol one more time every time. In the end, I'm no contributor to TS core and I definitely don't want to deal myself with possible complexity of this sugar.

It's not a dealbreaker guys. Also it takes courage and wisdom to spot request as too complex to implement (with outcome increasing overall code complexity), so hats off to you guys developing TypeScript and thank you for your work!

(p.s. came here to learn why not - in the end changed my mind and will deal with this._property = ...
and will be still happy)

**Comment by Maniae** [comment]
@idchlife The `set` property goal isn't to save developers one underscore when assigning value. You also might want to add some logic in the `set` process.

**Comment by FSM1** [comment]
Please reopen this issue. This is clearly a highly desired feature.

**Comment by Norby125** [comment]
+1 I would love to see this reconsidered.

**Comment by grosch** [comment]
Very strange your rationale is "this would make things confusing" when many other languages already do this.

**Comment by Zodiase** [comment]
Since all the 👍s are not doing much I'd just spam with one more 👍 comment.

**Comment by CreepyGnome** [comment]
This would be syntactic sugar which helps make things more readable and quick to implement. Lots of languages have syntactic sugar including Typescript. There doesn't seem to be any real good reasoning not to do it and too complex is never a good reason not to do something, it's more of a lazy cop-out.  Solving problems that are too complex is one of the reasons most people are drawn to software engineering.  So take this complex problem and solve it in a nice not so complex way. Will it require refactoring and risky changes? Maybe but that is how things get done.

**Comment by Kevin-at-Navigram** [comment]
Still desirable. Other languages have no trouble supporting different visibility for get- and set- functionality. To add injury to insult, Typescript is made by Microsoft, just like C#, but the latter fully supports it:

```
public string myPropertyWithLimitedAccess { get; private set; }   
```

Look at that. Beautifully readable, and totally nothing to complex.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties

There is little reason not to support:

```
private _myProperty: string;

public get myProperty(): string { return this._myProperty; }
private set myProperty(value: string) { this._myProperty = value; }
```

Bonus points in weird design decisions, because Typescript was primarily made for .NET developers to get more accustomed to working with the browser front-end.

**Comment by TomOeser** [comment]
+1 
Typescript would definitely benefit from this functionality!

**Comment by LevYas** [comment]
TypeScript is great even after C#, but this pointless limitation irritating me quite often. Please reopen it.

**Comment by Ace5273** [comment]
I think the only problem is with ts doing its job super well, but in this circumstance, it's just annoying so tried my way around ts and I think I've got it.

```
//@ts-ignore 
public static get IsOpenSession() : boolean{ 
//  Some code
}

//@ts-ignore 
private static set IsOpenSession(value: boolean){ 
// Some code
}
```

simple yet annoying 👍

**Comment by mcpiroman** [comment]
> would increase the complexity of the language

Is this really this complex comparing to things like `readonly [P in keyof T]: T[P]`?

**Comment by tomdaniel0x01** [comment]
Bump. Everyone want's this feature.

**Comment by snarfblam** [comment]
> > would increase the complexity of the language
> 
> Is this really this complex comparing to things like `readonly [P in keyof T]: T[P]`?

The complexity comes into play with the interaction of other language features. Unfortunately I can't find it, but IIRC RyanCavanaugh gave an example where this feature could allow you to set up and then violate invariants via inheritance using a class expression. Just because it's easy to write a certain declaration, that doesn't mean it'll always be easy to reason about how it affects things.

The question is which problems will a given feature solve, and which will it create. The former is easy to answer, the latter can surprisingly hard to answer. Unfortunately the TS team sometimes seems to respond with "OMG complexity" instead of illustrating the problem. (To be fair, the more time they spend answering query x for the nth time, the less time they have to develop.)

I don't 100% agree with the notion of "shouldn't the community get to decide", because if there is consensus among the experts actually developing the language, that should tell you something. But I think with something as requested as this, a thoughtful explanation from the team on what the trade-off is and why they're against it isn't too much to ask for.

And personally, I think the trade-off is 1000% worth it in this case. But if I can't be bothered to spec it out and make it work, I suppose I don't have too much right to complain.

**Comment by yvbeek** [comment]
Would this be a good time to revisit this issue?

The `readonly` modifier is great, but there are many cases where you want to be able to change the value inside of the class and still have everything be read-only on the outside.

I often choose not to make the property `readonly` because I don't like the noise that a private backing field + the properties add.

It would be great if there was some syntactic sugar to do this for you. Something like:

```typescript
// Option 1: C# style
public name: string { get; private set; }

// Option 2: Swift style
private(set) name: string

// Option 3: Swift struct-style
public readonly name: string

mutating changeName(name: string) {
  this.name = name
}

// Option 4: New keyword
public frozen name1: string
public readonly name2: string
```

I like option 2, imo it would fit well into the TypeScript language.

With option 4, `frozen` can only be set in the constructor, `readonly` can be set inside of this class, not by any external classes or classes inheriting from this class.

**Comment by trusktr** [comment]
For reference, @yvbeek's ideas on more flexible modifiers is better suited for the discussion over in https://github.com/microsoft/TypeScript/issues/37487.

This issues is specifically for getters and setters, and has a super high number of upvotes! I think it would be useful to give getters and setters differing access modifiers (and we can update the existing set of modifiers over in #37487 if the TypeScript team ever decides it to be an acceptable thing to move forward with)

**Comment by tomdaniel0x01** [comment]
> I don't 100% agree with the notion of "shouldn't the community get to decide", because if there is consensus among the experts actually developing the language, that should tell you something. But I think with something as requested as this, a thoughtful explanation from the team on what the trade-off is and why they're against it isn't too much to ask for.

Not to clog this thread with an irrelevant comment, but I think you have revealed a core principle for the way government should operate.

**Comment by CatalinMustata** [comment]
Not having this feature is a real pain for me and (among other things) made me switch from TS/NodeJS to something more... type safe. It's all fine and dandy, but when you're working on a complex project with a *a lot* of data structures (deeply nested many times) and you cannot model the data properly, you get the feeling that this is not a language "for big boys".

In my particular case, I want the property to be readonly, but modifiable from the inside... and also serialized to JSON. Too many hoops to jump through.

**Comment by Dazco** [comment]
I recently just switched to typescript and I've been loving the language. I'm really bummed out that this practical feature that exists in other languages hasn't been implemented here. Please do reconsider adding it in a future release regardless of how much complexity you think it might add to the language.

**Comment by ScreamZ** [comment]
I achieved something similar to c# with getters that way:
```ts
export class I18nService {
  private static ref: I18nService;

  public static get instance(): I18nService {
    if (!I18nService.ref) {
      I18nService.ref = new I18nService();
    }

    return I18nService.ref;
  }
}
```

**Comment by trusktr** [comment]
Type errors like the following are easy to understand and are not complicated:

```
Property 'foo' is writable only in protected scope within class 'Blah' and its subclasses.
```

or

```
Property 'foo' is readable only in protected scope within class 'Blah' and its subclasses.
```

etc, and similar with `private`.

That's honestly not complicated.

**Comment by danielrentz** [comment]
BTW, I stumbled over this problem too, and I am using this kind of "hack" in the meantime:

```typescript
// somewhere.ts
declare global {
    type Writable = { -readonly [P in keyof T]: T[P]; }
}

// example.ts
class Example {

    public readonly prop: number;

    public doSomething(n: number): void {
        (this as Writable<this>).prop =n;
    }
}
```
Technically, this could be used everywhere, but using this workaround should be restricted to code inside class methods.

**Comment by snarfblam** [comment]
> BTW, I stumbled over this problem too, and I am using this kind of "hack" in the meantime:

I've played with this idea myself, but the problem is you have no way to distinguish between "public read-only" and truly read-only properties. This makes it not very suitable as a general-purpose solution.

**Comment by RyanCavanaugh** [comment]
There's a [Playground build for PR 42425](https://www.staging-typescript.org/play?ts=4.3.0-pr-42425-10) available for trying out and I'd appreciate any feedback on how it meets folks' needs here. Thanks! See comments in #42425 on how this works.

**Comment by legistek** [comment]
"Adding new access modifiers to enable private_set/public_get would increase the complexity of the language and the learning curve, and the value gained from this would not match the added complexity."

Left hand (Typescript team), meet right hand (.NET team). Did you know you're part of the same company and have a lot of the same users?

**Comment by snarfblam** [comment]
This is nice. I hope it's a step toward asymmetric accessibility of data properties.

**Comment by legistek** [comment]
This is great to see. Apologies my prior comment was directed to the disnissive way this issue was summarily closed by the former team manager years ago. @RyanCavanaugh I'm glad to see typescript gaining more parity with languages like C#. Great work.

**Comment by jiriKralovec** [comment]
> The recommendation here is to use a public readonly getter and a private/protected backing field.
> 
> Accessors are symetrical with properties in the type system. anything we do will need to be manifested in the type and expressible on properties. Adding new access modifiers to enable private_set/public_get would increase the complexity of the language and the learning curve, and the value gained from this would not match the added complexity.

I am sorry but this is a rubbish reason. Typescript includes features like private constructors, yet you say that the complexity of different getters/setters is too high and steepens the learning curve. Like somebody mentioned here, `it is hard to make` should not be an argument against a standard in most modern programming languages. Your suggestion to use private/protected is not really helpful either - it makes the code less readable and it adds complexity to it.

**Comment by RiftLurker** [comment]
> I am sorry but this is a rubbish reason. Typescript includes features like private constructors, yet you say that the complexity of different getters/setters is too high and steepens the learning curve. Like somebody mentioned here, it is hard to make should not be an argument against a standard in most modern programming languages. Your suggestion to use private/protected is not really helpful either - it makes the code less readable and it adds complexity to it.

You might want to consider reading more than the first few comments. The one you quoted was written over 5 years ago, several months before even the addition of private constructors (TypeScript 2.0) you're talking about.
TypeScript has gone a long way since then, see the comments and linked commit (which landed in 4.3) right above.

**Comment by jiriKralovec** [comment]
@RiftLurker I know that, but that does not change the point. I did read the thread, and I do know the changes are in 4.3.
